### PR TITLE
Virtual pipeline parallel support for MegatronGPTSFTModel

### PR DESCRIPTION
**Pull request by vysarge**
# What does this PR do ?

Enables running MegatronGPTSFTModel with virtual pipeline parallel when peft_scheme=null.

**Collection**: NLP (language_modeling)

# Changelog 
- Alters word embedding initialization and passing of several parameters in MegatronGPTSFTModel to accommodate virtual pipeline parallelism when self.use_peft is false
- Adds a guard in `MegatronGPTSFTModel.__init__` to raise a ValueError if self.use_peft is true
- Corrects the TP group initialization check in MegatronGPTSFTModel to account for both TE and MCore flags

# Usage
Run a GPT model with SFT and set `virtual_pipeline_model_parallel_size`, such as:

```
python examples/nlp/language_modeling/tuning/megatron_gpt_peft_tuning.py \
++model.virtual_pipeline_model_parallel_size=10 \
...
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 1]
```diff
@@ -89,10 +89,8 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
...
 
-        if self.cfg.get('megatron_amp_O2', False):
-            base_module = self.model.module
-        else:
-            base_module = self.model
+        if self.use_peft and self.cfg.get('virtual_pipeline_model_parallel_size', None):
+            raise ValueError('Virtual pipeline model parallel is not supported when using PEFT')
 
         self._reset_activation_checkpointing_args()
         self._reset_sequence_parallelism_args()
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:96]: I guess we don't need to do this in this PR but we need to remove this from constructor or SP will not work with SFT.

Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:96]: Handled in https://github.com/NVIDIA/NeMo/pull/7963

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 2]
```diff
@@ -193,15 +191,19 @@ def setup(self, stage=None):
...
+                for index, module in enumerate(self.get_gpt_module_list()):
+                    if parallel_state.get_virtual_pipeline_model_parallel_world_size() is not None:
+                        parallel_state.set_virtual_pipeline_model_parallel_rank(index)
+                    sync_embeddings = (
+                        module.initialize_last_stage_with_word_embeddings
+                        if self.mcore_gpt
+                        else module.sync_initial_word_embeddings
+                    )
+                    sync_embeddings()
+                if parallel_state.get_virtual_pipeline_model_parallel_world_size() is not None:
```
[the remark this answers is not in the document]
Comment (vysarge) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py]: Done.  Can rename the function easily if we'd like, let me know if something else would better conform to naming convention.

**Comment by erhoo82** [comment]
jenkins

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 3]
```diff
@@ -349,16 +339,17 @@ def fwd_bwd_step(self, dataloader_iter, batch_idx, forward_only):
             grad_sync_func = self.reduce_overlap_gradients
             param_sync_func = self.sync_overlap_parameters
 
-        self.model.config.no_sync_func = no_sync_func
-        self.model.config.grad_sync_func = grad_sync_func
-        self.model.config.param_sync_func = param_sync_func
+        for module in self.get_gpt_module_list():
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:342]: this method's name changed in a recent PR

**Comment by ericharper** [comment]
jenkins

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**